Swap scene & ViewModel refactored

gemdev111 opened the pull request:
Integrated viewmodifier for DBObservering to viewmodel exectly Fixed issues where latests fetching quotes canceled Improved logic for swap
Cleanup
Moved actions logic to viewmodel for easy future tests implementation 
Migrated data provider fetching from viewmodel to different use case
Other minor updatees

closes #489

gemdev111 committed:
Swap scene & ViewModel refactored
Integrated viewmodifier for DBObservering to viewmodel exectly
Fixed issues where latests fetching quotes canceled
Improved logic for swap
Cleanup
Moved actions logic to viewmodel for easy future tests implementation
Migrated data provider fetching from viewmodel to different use case

gemcoder21 commented on Gem/Swap/Navigation/SwapNavigationView.swift:
```diff
@@ -4,79 +4,74 @@ import SwiftUI
-            )
-        }
+            }
+            .sheet(item: $model.presentedSheet) {
```
```suggestion
            .sheet(item: $model.isPresentedInfoSheet) {
```

gemcoder21 commented on Gem/Swap/Scenes/SwapScene.swift:
```diff
@@ -63,38 +32,39 @@ struct SwapScene: View {
+        .onChange(of: model.focusField, onChangeFocus)
         .onReceive(updateQuoteTimer) { _ in // TODO: - create a view modifier with a timer
-            fetch()
+            model.triggerFetch()
```
```suggestion
            model.fetch()
```

gemcoder21 commented on Gem/Swap/Scenes/SwapScene.swift:
```diff
@@ -145,20 +114,20 @@ extension SwapScene {
                     model: model.swapTokenModel(from: toAsset, type: .receive(chains: [], assetIds: [])),
                     text: $model.toValue,
-                    showLoading: model.showToValueLoading(),
+                    showLoading: model.showToValueLoading,
```
```suggestion
                    showLoading: model.isLoading,
```

gemcoder21 commented on Gem/Swap/Types/SwapState.swift:
```diff
@@ -2,20 +2,26 @@
         fetch: SwapFetchState = .idle,
         availability: StateViewType<[SwapQuote]> = .noData,
-        getQuoteData: StateViewType<Bool> = .noData
+        finalSwapData: StateViewType<TransferData> = .noData
```
```suggestion
        swapData: StateViewType<TransferData> = .noData
```

DRadmir replied:
Mb swapTransferData?

gemcoder21 commented on Gem/Swap/Types/SwapValueFormatter.swift:
```diff
@@ -0,0 +1,29 @@
+import Primitives
+import BigInt
+
+struct SwapValueFormatter {
```
add tests?

DRadmir commented on Gem/Swap/Scenes/SwapScene.swift:
```diff
@@ -107,33 +77,32 @@ extension SwapScene {
     private var swapFromSectionView: some View {
         Section {
-            if let fromAsset {
+            if let fromAsset = model.fromAsset {
```
I suggest encapsulating this logic in `SwapTokenView`. We can make model optional and this will mean empty state for View. And then check it like this: `if let model { ... } else { ... }`

This method `model.swapTokenModel(type: .pay)` will return an optional `SwapTokenViewModel`

DRadmir commented on Gem/Swap/Types/SwapQuoteInput.swift:
```diff
@@ -2,15 +2,41 @@
 
 extension SwapQuoteInput: Identifiable {
-    var id: String {"\(fromAsset?.id ?? "")_\(toAsset?.id ?? "")_\(amount)" }
+    var id: String {"\(fromAsset.id)_\(toAsset.id)_\(amount)" }
```
Not must have, just syntaxes sugar suggestion

```suggestion
var id: String { [fromAsset?.id ?? "", toAsset?.id ?? "", amount].joined(separator: "_") }
```

or

```suggestion
var id: String { [fromAsset?.id, toAsset?.id, amount].compactMap { $0 }.joined(separator: "_") }
```

DRadmir commented on Gem/Swap/ViewModels/SwapDataProvider.swift:
```diff
@@ -0,0 +1,114 @@
+    private let keystore: any Keystore
+    private let swapService: SwapService
+
+    init(keystore: any Keystore, swapService: SwapService) {
```
Should be public init? Or SwapDataProvider is internal struct?

gemdev111 replied:
it's inside gem target, If it's inside gem we don't need a public. For now only swap requires this, so if we migrated to  Swap package, guess it also should be internal

DRadmir commented on Gem/Swap/ViewModels/SwapSceneViewModel.swift:
```diff
@@ -0,0 +1,409 @@
+    }
+
+    func onChangeSwapQuoute(_ _: SwapQuote?, _ newQuote: SwapQuote?) {
+        guard let newQuote, let toAsset = toAsset else { return }
```
```suggestion
        guard let newQuote, let toAsset else { return }
```

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
minor refactoring, naming updates

gemdev111 committed:
Updates in SwapTokenView init

gemdev111 committed:
Fixed Swap tests by adding missed package
Migrated some types from gem swap target to Swap target
Added SwapValueFormatter tests
